File: trade/src/main/java/com/ez08/trade/ui/view/TradeMarketView.java
package com.ez08.trade.ui.view;

import android.content.Context;
import android.content.DialogInterface;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.util.AttributeSet;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.ez08.trade.R;
import com.ez08.trade.tools.CommonUtils;
import com.ez08.trade.tools.DialogUtils;
import com.ez08.trade.tools.MathUtils;
import com.ez08.trade.tools.YiChuangUtils;
import com.ez08.trade.ui.trade.ITradeView;
import com.ez08.trade.ui.trade.OptionsDelegate;
import com.ez08.trade.ui.trade.entity.TradeStockEntity;

public class TradeMarketView extends LinearLayout implements ITradeView, View.OnClickListener {

    Context context;
    ImageView plusNumIv;
    ImageView reduceNumIv;
    EditText inputCode;
    EditText editText2;

    RelativeLayout layout2;
    RelativeLayout layoutQuote;
    TextView full;
    TextView half;
    TextView one_three;
    TextView one_fourth;
    Button submitBtn;

    TextView maxText;
    TextView quoteWay;

    boolean bsflag = true;
    String quoteType = "市价委托";

    public TradeMarketView(Context context, AttributeSet attrs) {
        super(context, attrs);
        this.context = context;
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
        init();
    }

    private void init() {
        plusNumIv = findViewById(R.id.right_plus_num);
        reduceNumIv = findViewById(R.id.left_reduce_num);
        editText2 = findViewById(R.id.trade_entrust_num);
        inputCode = findViewById(R.id.input_code);

        layout2 = findViewById(R.id.layout2);
        layoutQuote = findViewById(R.id.layout_quote);
        layoutQuote.setOnClickListener(this);
        full = findViewById(R.id.full);
        half = findViewById(R.id.half);
        one_three = findViewById(R.id.one_three);
        one_fourth = findViewById(R.id.one_fourth);

        full.setOnClickListener(this);
        half.setOnClickListener(this);
        one_three.setOnClickListener(this);
        one_fourth.setOnClickListener(this);

        submitBtn = findViewById(R.id.submit);
        submitBtn.setOnClickListener(this);
        maxText = findViewById(R.id.available_num);
        quoteWay = findViewById(R.id.quote_way);

        change();

        inputCode.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
            }

            @Override
            public void afterTextChanged(Editable s) {
                if (inputCode.getText().length() == 6) {
                    delegate.search(inputCode.getText().toString());
                }
            }
        });

        inputCode.setOnFocusChangeListener(new OnFocusChangeListener() {
            @Override
            public void onFocusChange(View v, boolean hasFocus) {
                Log.e("onFocusChange", "" + hasFocus);
                if (hasFocus) {
                    inputCode.setText("");
                }
            }
        });

        plusNumIv.setOnClickListener(this);
        reduceNumIv.setOnClickListener(this);
    }

    public void change() {
        if (bsflag) {
            editText2.setBackgroundResource(R.drawable.trade_input_white_bg);
            plusNumIv.setColorFilter(getResources().getColor(R.color.trade_red));
            reduceNumIv.setColorFilter(getResources().getColor(R.color.trade_red));
            inputCode.setBackgroundResource(R.drawable.trade_input_bg);
            layout2.setBackgroundResource(R.drawable.trade_input_bg);
            layoutQuote.setBackgroundResource(R.drawable.trade_input_bg);
            full.setBackgroundResource(R.drawable.trade_input_bg);
            half.setBackgroundResource(R.drawable.trade_input_bg);
            one_three.setBackgroundResource(R.drawable.trade_input_bg);
            one_fourth.setBackgroundResource(R.drawable.trade_input_bg);
            submitBtn.setBackgroundResource(R.drawable.trade_red_corner_full);
            submitBtn.setText("立即买入");
        } else {
            editText2.setBackgroundResource(R.drawable.trade_input_white_bg_blue);
            plusNumIv.setColorFilter(getResources().getColor(R.color.trade_blue));
            reduceNumIv.setColorFilter(getResources().getColor(R.color.trade_blue));
            inputCode.setBackgroundResource(R.drawable.trade_input_bg_blue);
            layout2.setBackgroundResource(R.drawable.trade_input_bg_blue);
            layoutQuote.setBackgroundResource(R.drawable.trade_input_bg_blue);
            full.setBackgroundResource(R.drawable.trade_input_bg_blue);
            half.setBackgroundResource(R.drawable.trade_input_bg_blue);
            one_three.setBackgroundResource(R.drawable.trade_input_bg_blue);
            one_fourth.setBackgroundResource(R.drawable.trade_input_bg_blue);
            submitBtn.setBackgroundResource(R.drawable.trade_blue_corner_full);
            submitBtn.setText("立即卖出");
        }
    }

    @Override
    public void onClick(View v) {
        if (v == plusNumIv) {
            String e1 = editText2.getText().toString();
            if (TextUtils.isEmpty(e1)) {
                e1 = "0";
            }
            int t1 = Integer.parseInt(e1);
            t1 = t1 + 100;
            editText2.setText(t1 + "");
        } else if (v == reduceNumIv) {
            String e1 = editText2.getText().toString();
            if (TextUtils.isEmpty(e1)) {
                return;
            }
            int t1 = Integer.parseInt(e1);
            t1 = t1 - 100;
            if (t1 < 0) {
                t1 = 0;
            }
            editText2.setText(t1 + "");
        } else if (v == full) {
            editText2.setText(MathUtils.save100(maxValue));
        } else if (v == half) {
            editText2.setText(MathUtils.save100(maxValue / 2));
        } else if (v == one_three) {
            editText2.setText(MathUtils.save100(maxValue / 3));
        } else if (v == one_fourth) {
            editText2.setText(MathUtils.save100(maxValue / 4));
        } else if (v == submitBtn) {
            String stockCode = inputCode.getText().toString().trim();
            String num = editText2.getText().toString().trim();
            if (isInvalid(stockCode,num)) {
                delegate.submit(entity.stkcode, price, 0,num, quoteWay.getText().toString());
            }
        } else if (v == layoutQuote) {
            if (entity == null || entity.market == null) {
                CommonUtils.show(context, "请输入股票代码");
                return;
            }

            final String[] select;
            if (entity.market.equals("0")) {
                select = YiChuangUtils.szQuoteType;
            } else if (entity.market.equals("1")) {
                select = YiChuangUtils.shQuoteType;
            } else {
                select = new String[]{};
            }
            DialogUtils.showSelectDialog(context, select, new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialog, int which) {
                    quoteWay.setText(select[which]);
                }
            });
        }
    }

    TradeStockEntity entity;
    int maxValue = 0;
    String price;

    public void setStockEntity(TradeStockEntity entity) {
        this.entity = entity;
        inputCode.setText(entity.stkcode + "    " + entity.stkname);
        price = MathUtils.format2Num(entity.fNewest);
        delegate.getMax(entity.stkcode, price);

        this.requestFocus();
        this.requestFocusFromTouch();
    }

    @Override
    public void setBsflag(boolean flag) {
        this.bsflag = flag;
        if (inputCode != null) {
            change();
        }
    }

    public void setMax(String max) {
        if (bsflag) {
            maxText.setText("最大可买 " + max + " 股");
        } else {
            maxText.setText("最大可卖 " + max + " 股");
        }
        maxValue = Integer.parseInt(max);
    }

    @Override
    public void setStockCode(String code) {
        inputCode.setText(code);
    }


    OptionsDelegate delegate;

    public void setDelegate(OptionsDelegate delegate) {
        this.delegate = delegate;

    }

    private boolean isInvalid(String stockCode, String num) {
        if (entity==null|| TextUtils.isEmpty(stockCode) || TextUtils.isEmpty(num)) {
            CommonUtils.show(context, R.string.trade_notice_msg);
            return false;
        }
        return true;

    }
}

File: app/src/main/java/com/god/kotlin/trade/TradeMarketView.kt
package com.god.kotlin.trade

import android.content.Context
import android.content.DialogInterface
import android.text.TextUtils
import android.widget.RelativeLayout
import androidx.core.content.ContextCompat
import com.ez08.trade.tools.DialogUtils
import com.god.kotlin.R
import com.god.kotlin.data.entity.Avail
import com.god.kotlin.data.entity.TradeStockEntity
import com.god.kotlin.user.UserHelper
import com.god.kotlin.util.*
import com.god.kotlin.widget.level1.Level
import kotlinx.android.synthetic.main.view_top.view.*
import kotlinx.android.synthetic.main.view_trade.view.*
import kotlinx.android.synthetic.main.view_trade_market.view.*
import kotlinx.android.synthetic.main.view_trade_market.view.available_num
import kotlinx.android.synthetic.main.view_trade_market.view.input_code
import kotlinx.android.synthetic.main.view_trade_market.view.level_view
import kotlinx.android.synthetic.main.view_trade_market.view.submit
import kotlinx.android.synthetic.main.view_trade_market.view.total_num
import kotlinx.android.synthetic.main.view_trade_ratio.view.*

class TradeMarketView(context: Context?) : RelativeLayout(context), ITradeView {

    private var list = mutableListOf<TradeStockEntity.Dang>()
//    private var adapter: LevelAdapter
    private lateinit var viewModel: SellViewModel
    private var direction: Boolean = true
    private var _data: TradeStockEntity? = null
    private var maxValue: Int = 0
    private val levelView: Level

    init {
        context!!.inflate(R.layout.view_trade_market, this)
//        adapter = LevelAdapter(list, context)
//        level_view.adapter = adapter
//        level_view.setOnItemClickListener { parent, view, position, id ->
//            val prices = list[position]
////            price.text = prices.fPrice.format2()
//        }
        levelView = (level_view as Level)
        levelView.setOnItemClickListener(object : Level.OnItemClickListener {
            override fun onClick(position: Int) {
//                val prices = list[position]
//                price.text = prices.fPrice.format2()
            }
        })

        input_code.setOnSearchListener {
            viewModel.search(it)
        }

        layout_quote.setOnClickListener {
            _data?.let {
                val select: Array<String> = when (it.market) {
                    "0" -> szQuoteType
                    "1" -> shQuoteType
                    else -> arrayOf()
                }
                showSelectDialog(context, select,
                    DialogInterface.OnClickListener { _, which -> quote_way.text = select[which] })
            } ?: "请输入代码".toast(context)

        }

        full.setOnClickListener {
            total_num.text = maxValue.save100()
        }

        half.setOnClickListener {
            total_num.text = (maxValue / 2).save100()
        }

        one_three.setOnClickListener {
            total_num.text = (maxValue / 3).save100()
        }

        one_fourth.setOnClickListener {
            total_num.text = (maxValue / 4).save100()
        }

        submit.setOnClickListener {
            _data?.let {
                if(_data == null){
                    "无效的股票，请稍后再试".toast(context)
                    return@setOnClickListener
                }

                if(TextUtils.isEmpty(total_num.text)){
                    "请输入有效数量".toast(context)
                    return@setOnClickListener
                }

                val user = UserHelper.getUserByMarket(it.market)
                val postFlag = getTagByQuoteName(if (direction) "0B" else "0S", quote_way.text.toString())
                if(TextUtils.isEmpty(postFlag)){
                    "请选择报价方式".toast(context)
                    return@setOnClickListener
                }

                val option = if (direction) "买入" else "卖出"

                DialogUtils.showTwoButtonDialog(
                    context, option + "交易确认", "确定$option",
                    "操作类型：" + option + "\n" +
                            "股票代码：" + it.stkcode + "  " + it.stkname + "\n" +
                            "委托数量：" + total_num.text.toIntOrZero() + "\n" +
                            "委托方式：" + quote_way.text.toString() +  postFlag + "\n" +
                            "股东代码：" + user.secuid
                ) { _, _ ->
                    viewModel.transaction(
                        it.market, it.stkcode, user.secuid, user.fundid,
                        it.fixprice, total_num.text.toInt(), postFlag
                    )
                }
            }
        }

    }

    override fun initData(flag: Boolean, vm: SellViewModel) {
        viewModel = vm
        direction = flag

        if (direction) {
            input_code.setBackgroundResource(R.drawable.trade_input_bg)
            total_num.setColor(R.color.trade_red)
            submit.setBackgroundResource(R.drawable.trade_red_corner_full)
            layout_quote.setBackgroundResource(R.drawable.trade_input_bg)
            submit.text = "立即买入"
        } else {
            input_code.setBackgroundResource(R.drawable.trade_input_bg_blue)
            total_num.setColor(R.color.trade_blue)
            layout_quote.setBackgroundResource(R.drawable.trade_input_bg_blue)
            submit.setBackgroundResource(R.drawable.trade_blue_corner_full)
            submit.text = "立即卖出"
        }
    }

    override fun setStockCode(code: String) {
        input_code.search(code)
    }

    override fun setAvailable(avail: Avail) {
        if (avail.direction == direction) {
            if (avail.direction) {
                maxValue = avail.num
                available_num.text = "可买${maxValue}股"
            } else {
                maxValue = avail.num
                available_num.text = "可卖${maxValue}股"
            }
        }
    }

    override fun setData(data: TradeStockEntity) {
        val user = UserHelper.getUserByMarket(data.market)

        data.let {
            _data = it
            input_code.setData(it.stkcode, it.stkname)

            limit_up_price.text = it.maxrisevalue.format2()
            limit_up_price.setTextColor(ContextCompat.getColorStateList(context, R.color.trade_red))
            limit_down_price.text = it.maxdownvalue.format2()
            limit_down_price.setTextColor(ContextCompat.getColorStateList(context, R.color.trade_green))
            viewModel.getAvailable(
                data.market, user.secuid, user.fundid,
                it.stkcode, it.maxrisevalue, direction
            )
        }
    }

    override fun updateHQ(data: TradeStockEntity) {
        data.let {
            newest_price.text = it.fNewest.format2()
            newest_price.setTextColor(ContextCompat.getColorStateList(context, getPriceColor(it.fNewest, it.fOpen)))
            last_price.text = it.fLastClose.format2()

            list.clear()

            val temp = mutableListOf<TradeStockEntity.Dang>()
            temp.addAll(data.ask)
            temp.reverse()

            list.addAll(temp)
            list.addAll(data.bid)
//            adapter.setOpenPrice(it.fOpen)
//            adapter.notifyDataSetChanged()
            levelView.setOpenPrice(it.fOpen)
            levelView.update(list)
        }
    }
}

